### PR TITLE
feat: saga machine example

### PR DESCRIPTION
**Pull request by thantos**
Depends on https://github.com/functionless/functionless/pull/361 (API Gateway Template Formatting)

Based on: https://github.com/cdk-patterns/serverless/tree/main/the-saga-stepfunction

1. Book a hotel
2. Book a flight
3. Pay
4. Confirm flight
5. Confirm hotel

If any steps fail, undo any already executed actions. 

ex: 

If the payment fails, revert the flight booking and the hotel booking.

 Two Constructs;
1. Saga - 1:1 with the source stack. 1 step function and 8 lambda functions for each step.
2. Optimized Saga - 1 step function, 1 lambda function for creating ID hashes.

### CODE REVIEW

**Comment by thantos** [comment]
Random Metrics:

1. Original Code - 9 files (8 lambda, 1 step function construct) - 650 lines of code
2. Saga in Functionless - 1 file (construct - 8 lambdas 1 step function) - ~450 lines of code - 6.5 seconds cold start, 2.5 seconds after cold start runtime avg 
3. Optimized Saga in Functionless - 1 file (construct - 1 lambda 1 step function) - ~200 lines of code - 2.5 second runtime avg

File: packages/sagaFunction/README.md [hunk 1]
```diff
@@ -0,0 +1 @@
+# replace this
```
Comment (sam-goodwin) [inline packages/sagaFunction/README.md]: I like the content in the original pattern. Should we have the same?

**Review by sam-goodwin — COMMENTED** [review]
Beautiful

**Review by sam-goodwin — APPROVED** [review]

**Comment by sam-goodwin** [comment]
These seem hard to read, do we know the source of these inefficiencies? Both in terms of naming and excessive states?

image
image
image

Equivalent when hand-written without Functionless
image

**Pull request merged**